### PR TITLE
XCM: Implement a blocking barrier

### LINKED ISSUE #4813: XCM: Blocking Barrier
Sometimes it can be useful to block all but a particular type of XCM from being executed. This would allow for that by:

- Introducing a proper Error type for the `ShouldExecute::should_execute`. This type would include the `Suspended` variant meaning "does not pass this barrier yet, but it might pass in a later block" .
- Introducing a new struct `RespectSuspension` impl for `ShouldExecute` in PalletXcm which takes an `Inner: ShouldExecute` and which returns `Suspended` if a state value `SuspendFlag` is set to `true` and returns the result of its `Inner::should_execute` otherwise.

This can work with more broad queue suspensions to ensure that only acceptable messages are processed under certain circumstances.

### PR DESCRIPTION
**Pull request by KiChjang**
Rebased version of #5035.

Fixes #4813.

cumulus companion: paritytech/cumulus#2174

### CODE REVIEW

File: xcm/pallet-xcm/src/lib.rs [hunk 1]
```diff
@@ -1090,6 +1096,18 @@ pub mod pallet {
...
 		}
+
+		/// Set or unset the global suspension state of the XCM executor.
+		///
+		/// - `origin`: Must be Root.
+		/// - `suspended`: `true` to suspend, `false` to resume.
+		#[pallet::call_index(10)]
+		#[pallet::weight(100_000_000u64)]
+		pub fn force_suspension(origin: OriginFor<T>, suspended: bool) -> DispatchResult {
+			ensure_root(origin)?;
```
Comment (gavofyork) [inline xcm/pallet-xcm/src/lib.rs:1107]: Should probably be a configuration item.

**Comment by gavofyork** [comment]
I don't think this actually does what it's meant to do.

> This type would include the `Suspended` variant meaning "does not pass this barrier yet, but it might pass in a later block" .

This meaning was not implemented. It will need to be integrated properly into the Message Queue systems (i.e. pallet message queue), so they stop executing messages from any origin which returns `Suspended` until some time in the future (e.g. the next block).

**Review by gavofyork — CHANGES_REQUESTED** [review]
I don't think this actually does what it's meant to do.

> This type would include the `Suspended` variant meaning "does not pass this barrier yet, but it might pass in a later block" .

This meaning was not implemented. It will need to be integrated properly into the Message Queue systems (i.e. pallet message queue), so they stop executing messages from any origin which returns `Suspended` until some time in the future (e.g. the next block).

As it stands it will do the same as it would for any other non-passing barriers: just drop the messages. Definitely not the intended behaviour.

**Comment by girazoki** [comment]
I observe the PR includes a db read made by a barrier. I am assuming this could produce an unpaid db read, is it not that much of an issue because it can at most spam one db read per block? (I am assuming following messages processed in the same block will have the db read cached?)

**Comment by gavofyork** [comment]
This should help https://github.com/paritytech/substrate/pull/13424

**Comment by gavofyork** [comment]
@KiChjang you going to get this one over the line now? With the `Yield` error it should be possible to implement properly.

**Comment by KiChjang** [comment]
Wouldn't this be blocked on #6271 before we can actually use the queues here?

**Comment by gavofyork** [comment]
possibly, but it doesn't matter - the API is in.

**Comment by KiChjang** [comment]
Will reopen this PR with some changes.

**Pull request closed without merging**